scripts/hud/jump-timing.ts

```diff
@@ -21,31 +20,61 @@ class JumpTimingHandler {
 		lateBarProgress: $<Panel>('#JumpLateBar_Left')
 	};
 
+	perfWindow: number;
+	range: number;
+	bufferedJumpingEnabled: boolean;
+
 	constructor() {
 		RegisterHUDPanelForGamemode({
 			gamemodes: GamemodeCategories.get(GamemodeCategory.BHOP), // TODO: Add stamina bhop game mode
+			onLoad: () => this.onMapInit(),
+			events: [
+				{
+					event: 'JumpTimingSettingChanged',
+					callback: () => this.updateSettings()
+				}
+			],
 			handledEvents: [
 				{
 					event: 'JumpTimingDataUpdate',
 					panel: this.panels.container,
-					callback: (jumpTiming, maxTimingPenalty) => this.onJumpTimingUpdate(jumpTiming, maxTimingPenalty)
+					callback: (jumpTiming) => this.onJumpTimingUpdate(jumpTiming)
 				}
 			]
 		});
 	}
 
-	onJumpTimingUpdate(jumpTiming: number, maxTimingPenalty: number) {
-		// Blend between yellow and orange for early timing, and orange and red for late timing since you lose more speed on the ground from friction and stamina
-		const earlyTimingRatio = jumpTiming < 0 ? -jumpTiming / maxTimingPenalty : 0;
-		this.panels.earlyBar.value = earlyTimingRatio;
-		this.panels.earlyLabel.text = jumpTiming < 0 ? -jumpTiming : 0;
-		const colorLerp = (Math.abs(jumpTiming) - 1) / (maxTimingPenalty - 1);
-		const earlyTimingColor = rgbaTupleLerp(Colors.YELLOW, Colors.ORANGE, colorLerp);
-		this.panels.earlyBarProgress.style.backgroundColor = tupleToRgbaString(earlyTimingColor);
+	onMapInit() {
+		this.updateSettings();
+	}
+
+	updateSettings() {
+		this.range = GameInterfaceAPI.GetSettingInt('mom_hud_jump_timing_range');
+		this.perfWindow = GameInterfaceAPI.GetSettingInt('mom_mv_perfect_jump_window');
+		this.bufferedJumpingEnabled = GameInterfaceAPI.GetSettingInt('mom_mv_autohop_mode') === 2;
+
+		this.panels.earlyLabel.visible = this.bufferedJumpingEnabled;
+		this.panels.earlyBar.visible = this.bufferedJumpingEnabled;
+		this.panels.lateBar.style.width = this.bufferedJumpingEnabled ? '50%' : '100%';
+	}
+
+	onJumpTimingUpdate(jumpTiming: number) {
+		jumpTiming = Math.max(Math.min(jumpTiming, this.range), -this.range);
+
+		if (this.bufferedJumpingEnabled) {
+			const earlyTimingRatio =
+				jumpTiming < 0
+					? Math.max(-jumpTiming - this.perfWindow, 0) / Math.max(this.range - this.perfWindow, 1)
+					: 0;
+			this.panels.earlyBar.value = earlyTimingRatio;
+			this.panels.earlyLabel.text = jumpTiming < 0 ? -jumpTiming : 0;
+			this.panels.earlyBarProgress.style.backgroundColor = tupleToRgbaString(Colors.ORANGE);
```

Comment:
I don't think this is particularly expensive, so feel free to ignore to keep the early/late code looking similar. If this is constant, it could be applied once on init, or in `updateSettings()`.